modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/schema/SqlSchemaManagerImpl.java

```diff
@@ -101,10 +109,12 @@ public SqlSchemaManagerImpl(CatalogManager catalogManager, CacheFactory factory,
     /** {@inheritDoc} */
     @Override
     public SchemaPlus schema(int catalogVersion) {
-        return schemaCache.get(
+        SchemaPlus schemaPlus = schemaCache.get(
                 catalogVersion,
                 version -> createRootSchema(catalogManager.catalog(version))
         );
+
+        return makeSchemaCopyWithStatisticSnapshot(schemaPlus, catalogVersion);
```

Comment:
I think this is not a good idea to re-create calcite's schema for each call. Could you please run some benchmarks to show that these change have does not affect performance?

Because with this change every caller is going to pay the price for statistics even when they do not need it (see fast plans)

You can move this code to `PrepareServiceImpl`  `prepareXXX` methods because statistics is not needed unless the SQL engine decides to run full optimisation.

Reply:
Unfortunately, we don't have now another way to implement it (or I don't know such way), due to schema required for execute every SQL and we haver a cache for schemas.
It's lightweight copying. Statistics not calculated every time, just look into the constructor of StatisticsSnapshot `new Lazy<>(statistic::getRowCount);` - it's lazy evaluation, just for cases when it really required.
Yes, I will provide benchmark results a little bit later.

Reply:
reworked